specification/cdn/resource-manager/Microsoft.Cdn/stable/2020-04-15/cdn.json

```diff
@@ -0,0 +1,5518 @@
+          "type": "string"
+        }
+      }
+    },
+    "OriginUpdateParameters": {
+      "type": "object",
+      "description": "Origin properties needed for origin update.",
+      "properties": {
+        "tags": {
+          "description": "Origin tags.",
+          "type": "object",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/OriginUpdatePropertiesParameters"
+        }
+      },
+      "x-ms-azure-resource": true
+    },
+    "OriginUpdatePropertiesParameters": {
+      "description": "The JSON object that contains the properties of the origin.",
+      "properties": {
+        "hostName": {
+          "description": "The address of the origin. Domain names, IPv4 addresses, and IPv6 addresses are supported.This should be unique across all origins in an endpoint.",
+          "type": "string"
+        },
+        "httpPort": {
+          "description": "The value of the HTTP port. Must be between 1 and 65535.",
+          "type": "integer",
+          "maximum": 65535,
+          "exclusiveMaximum": false,
+          "minimum": 1,
+          "exclusiveMinimum": false
+        },
+        "httpsPort": {
+          "description": "The value of the HTTPS port. Must be between 1 and 65535.",
+          "type": "integer",
+          "maximum": 65535,
+          "exclusiveMaximum": false,
+          "minimum": 1,
+          "exclusiveMinimum": false
+        },
+        "originHostHeader": {
+          "description": "The host header value sent to the origin with each request. If you leave this blank, the request hostname determines this value. Azure CDN origins, such as Web Apps, Blob Storage, and Cloud Services require this host header value to match the origin hostname by default. This overrides the host header defined at Endpoint",
+          "type": "string"
+        },
+        "priority": {
+          "description": "Priority of origin in given origin group for load balancing. Higher priorities will not be used for load balancing if any lower priority origin is healthy.Must be between 1 and 5",
+          "type": "integer",
+          "maximum": 5,
+          "minimum": 1
+        },
+        "weight": {
+          "description": "Weight of the origin in given origin group for load balancing. Must be between 1 and 1000",
+          "type": "integer",
+          "maximum": 1000,
+          "minimum": 1
+        },
+        "enabled": {
+          "description": "Origin is enabled for load balancing or not",
+          "type": "boolean"
+        },
+        "privateLinkAlias": {
+          "description": "The Alias of the Private Link resource. Populating this optional field indicates that this origin is 'Private'",
+          "type": "string"
+        },
+        "privateLinkResourceId": {
```

Comment:
Frontdoor seems to do the linking with just the alias.  Why differ in contract here?  Why do you need all 3 properties? Location definitely should not be needed here as you can get it from the resourceId if necessary.

Reply:
In addition to the alias, we also allow the combination of resource Id and location, since the location cannot be easily inferred from the resource id.

Currently with only the alias, our customers can only use Virtual Machines as backends. However, with the addition of the resource ID and location, we can now support other resource types hidden behind a private link (e.g. Storage, Web services, Keyvault). Since these resource types allow you to create a private link behind the scenes, while only providing the resource ID but not the alias.

Reply:
you can't trust the location the user gives you is accurate so you need to validate the resource is in that location.  The user shouldn't need to provide information that can easily be looked up and that they may get wrong.

Why do privateLinkServices only have aliases if they are linked to a VM?  Or are there other resource IDs that you would link to here?

Reply:
The location doesn't *have* to match the resource ID. We use the location to create a private endpoint in Frontdoor's VNET that in the region supplied by the customer. If the user provides a different region, there will be additional billings and latency as the request is sent across regions.

Private links can be created in different ways, one of which is through Add -> Private link and linking it to a load balancer, which only supports virtual machines as backends. If a PLS is created this way, then the owner can share both the resource ID (and location) or an alias.

Another way to create a PLS is to create it from under a resource. For example, if a user owns a storage resource, they can go to said resource and create a private endpoint which creates a Private Link behind scenes. However, the customer doesn't have access to that private link that was created unless they query an endpoint that returns the private links created for this resource  (e.g. `GET /subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Storage/storageAccounts/storageName/privateLinkResources?api-version=2019-04-01`). Since this endpoint returns the resource ID of the private links, we needed to add the ability to provide it through our API